src/twisted/internet/tcp.py

```diff
@@ -1038,42 +1285,62 @@ def doRead(self):
                         # connection, but we get told to try to accept()
                         # anyway.
                         continue
-                    elif e.args[0] in (EMFILE, ENOBUFS, ENFILE, ENOMEM, ECONNABORTED):
-                        # Linux gives EMFILE when a process is not allowed to
-                        # allocate any more file descriptors.  *BSD and Win32
-                        # give (WSA)ENOBUFS.  Linux can also give ENFILE if the
-                        # system is out of inodes, or ENOMEM if there is
-                        # insufficient memory to allocate a new dentry.
-                        # ECONNABORTED is documented as possible on all
-                        # relevant platforms (Linux, Windows, macOS, and the
-                        # BSDs) but occurs only on the BSDs.  It occurs when a
-                        # client sends a FIN or RST after the server sends a
-                        # SYN|ACK but before application code calls accept(2).
-                        # On Linux, calling accept(2) on such a listener
-                        # returns a connection that fails as though the it were
-                        # terminated after being fully established.  This
-                        # appears to be an implementation choice (see
-                        # inet_accept in inet/ipv4/af_inet.c).  On macOS,
-                        # such a listener is not considered readable, so
-                        # accept(2) will never be called.  Calling accept(2) on
-                        # such a listener, however, does not return at all.
+                    elif e.args[0] == EMFILE and self._reservedFD.available():
+                        # Linux gives EMFILE when a process is not
+                        # allowed to allocate any more file
+                        # descriptors.  *BSD and Win32 give
+                        # (WSA)ENOBUFS.
+                        log.msg(
+                            "EMFILE encountered;"
+                            " releasing reserved file descriptor.")
+                        self._reservedFD.release()
+                        continue
+                    elif e.args[0] in (
+                            EMFILE, ENOBUFS, ENFILE, ENOMEM, ECONNABORTED):
```

Comment:
Totally optional: It might be nice to hoist this tuple to module scope, along with the comment (which would then have a bit more horizontal space and be easier to read).